### PR TITLE
Build the run-time into libverilated.a ahead of time

### PR DESCRIPTION
**Pull request by gezalore**
This is a draft of building the run-time lib into an archive ahead of time. I wanted to run it past you to see if you would be happy with this before I do more work on it.

It's actually a surprisingly small patch that mostly works. Namely, the failures left in the test suite are kind of trivial and relate to selecting the debug/release variants of the run-time lib.

verilator_ext_tests and XiangShan build fine. OpenTitan needs fixing the support for overriding vl_stop, but that's quite easy in this model.

Changes to user interface for simple users (those that use the default makefile/cmake):

1. Ideally, selecting the debug build of the run-time library would have to change a bit. Currently we use -DVL_DEBUG=1 or -UVL_DEBUG. It would be unfortunate if we had to scan the CPPFLAGS to see which one to pick up, so maybe add an explicit make parameter VL_DEBUG=1/0 (CMake should be able to pick up the right one based on the CMake build type). (This is the cause of the remaining upstream test failures.

2. Because we build the run-time ahead of time, the SystemC library is kind of baked in. Although we only use the SystemC headers during build type, (and binaries during testing), it would still likely less problematic to require Verilator be built against the specific SystemC library the user intends to use. For this we should add a `--with-systemc` option to configure, which defaults to the system version (if any).

Changes for advanced users:

Instead of picking up the particular source files, link against the appropriate flavour of libverilator_rt that lives in the `lib/` directory.

The nice thing about using an archive is that overriding vl_stop/vl_finish and friends just works by putting them into separate object files during the library build (and using the correct link order, namely libverilator_rt after all user archives), and ensuring they are pulled in if the user provides them from archives themselves.

We could also just apply this to v5 if you prefer, due to the [hopefully minor] user perceptible changes.

### CODE REVIEW

**Comment by gezalore** [comment]
Note: I haven't moved any sources around from `include/`, but it's contents should move under `src/runtime` or similar if we do this.

**Comment by gezalore** [comment]
Also, V<model>_ALL.a now no longer contains the run-time, but that can be fixed if desirable

**Review by wsnyder — COMMENTED** [review]
>We could also just apply this to v5 if you prefer, due to the [hopefully minor] user perceptible changes.

I think we should do it against v5, as this will make big changes in how distro's .rpms are packaged.  Please don't move any source files as it will be hard to merge.


>maybe add an explicit make parameter VL_DEBUG=1/0 (CMake should be able to pick up the right one based on the CMake build type). (This is the cause of the remaining upstream test failures.

Seems reasonable

>we should add a --with-systemc option to configure, which defaults to the system version (if any).

I find this distasteful but don't have a better alternative.  I don't think many people run with different SystemC versions any longer (like we had to when it was a young library.)

>The nice thing about using an archive is that overriding vl_stop/vl_finish and friends just works by putting them into separate object files during the library build

I agree though am not sure this is what we want to encourage as the official API.  Also relying on link order is traditionally confusing to the inexperienced.


>Another note: Technically there are 8 flavours of the library: {single-threaded,multi-threaded}x{without-systemc,with-systemc}x{debug,optimized}. So in reality that is 4 flavours, without the debug/release distinction.

I was wondering if we should show it this way, e.g. _rt_{mt|st}_{cc|sc|{_dbg}.  Thinking about it further if we add more options the most "basic" should not need to have a name change, but single threaded vs multithreaded is important.  Also "libverilated" is by definition the runtime, so I think the _rt is not needed.  So I suggest naming libverilated_{mt|st}{_sc}{_dbg}.a

>Also, V_ALL.a now no longer contains the run-time, but that can be fixed if desirable

Seems fine, perhaps better?  But then downstream needs to know verilator.lib flavor

How do we detect mismatches between linking "wrong" version e.g. verilated non-mt against a --threads 0 model?

After this is otherwise working please make sure "make coverage" properly collects the library coverage. Should be easier as previously it had to look in each test.

Please move src/Makefile_obj_rt.in elsewhere, to wherever we think the runtime sources will be (src/rt?)  It probably should be Makefile_obj.mk, the missing .mk is historical and probably time to clean up.

What about gtkwave source files?

Maybe we should keep the sources in include, I realize this isn't standard but I hate having .cpp files far away from headers.  Also better if someone does want to combine all sources at build time versus using the library for maximum performance.

**Comment by gezalore** [comment]
> > We could also just apply this to v5 if you prefer, due to the [hopefully minor] user perceptible changes.
> 
> I think we should do it against v5, as this will make big changes in how distro's .rpms are packaged. Please don't move any source files as it will be hard to merge.

Agreed. I won't move any files until master is frozen.
 
> > maybe add an explicit make parameter VL_DEBUG=1/0 (CMake should be able to pick up the right one based on the CMake build type). (This is the cause of the remaining upstream test failures.
> 
> Seems reasonable
> 
> > we should add a --with-systemc option to configure, which defaults to the system version (if any).
> 
> I find this distasteful but don't have a better alternative. I don't think many people run with different SystemC versions any longer (like we had to when it was a young library.)

So, looking at the usage of SystemC in cpp files, there are very very few. I think we could make all the SystemC dependencies header only, that way we don't need SystemC at build time, so no need to bake it in. Might be a bit harder to maintain (though we should just add a test that proves we don't depend on it when building the run-time), but hopefully we won't have to add many more dependencies on SystemC. Thoughts?

(BTW I'm not even sure why Verilator (the compiler part) needs to be aware of the SystemC library, i.e.: why we need configure to figure it out?. I get it we use it for testing, but we could just do that in driver.pl...).

> > The nice thing about using an archive is that overriding vl_stop/vl_finish and friends just works by putting them into separate object files during the library build
> 
> I agree though am not sure this is what we want to encourage as the official API.

Agreed, my point was providing this for compatibility purposes. Though as always, I'm happy to break things if preferred :smiling_imp: 

> Also relying on link order is traditionally confusing to the inexperienced.

For the basic Verilator driven build, I will just make this work. If you roll your own build system, I suggest we assume you know what you are doing (or want to learn).
 
> > Another note: Technically there are 8 flavours of the library: {single-threaded,multi-threaded}x{without-systemc,with-systemc}x{debug,optimized}. So in reality that is 4 flavours, without the debug/release distinction.
> 
> I was wondering if we should show it this way, e.g. _rt_{mt|st}_{cc|sc|{_dbg}. Thinking about it further if we add more options the most "basic" should not need to have a name change, but single threaded vs multithreaded is important. Also "libverilated" is by definition the runtime, so I think the _rt is not needed. So I suggest naming libverilated_{mt|st}{_sc}{_dbg}.a

There should really be no circumstance under which we add a new flavour (and if we can get rid of the SystemC flavour as above, then even better). I am even temped to run some benchmarking to see if the st/mt distinction is still warranted (I am under the impression that those two are functionally equivalent for single threaded models, only speed differs).

As for libverilated, using an adjective as a proper noun tickles my brain the wrong way, but I can live with that :)
 
> > Also, V_ALL.a now no longer contains the run-time, but that can be fixed if desirable
> 
> Seems fine, perhaps better? But then downstream needs to know verilator.lib flavor
> 
> How do we detect mismatches between linking "wrong" version e.g. verilated non-mt against a --threads 0 model?

We add a symbol unique to each flavour, and make the generated models have a reference to the appropriate one, that way you get a link time undefined symbol error if linking the wrong one. (assuming we do end up having multiple flavours?)

> After this is otherwise working please make sure "make coverage" properly collects the library coverage. Should be easier as previously it had to look in each test.
> 
> Please move src/Makefile_obj_rt.in elsewhere, to wherever we think the runtime sources will be (src/rt?) It probably should be Makefile_obj.mk, the missing .mk is historical and probably time to clean up.
> 
> What about gtkwave source files?

They are all #included in cpp files and not built directly (and are part of libverilated). I don't particularly mind if they live next to the rest of the sources.

> Maybe we should keep the sources in include, I realize this isn't standard but I hate having .cpp files far away from headers. Also better if someone does want to combine all sources at build time versus using the library for maximum performance.

My intention was to move all current content of `include` under `src/runtime`, then as part of the build steps, symlink the public headers only under `include` (and copy them as part of the install step). There are headers in `include` that is none of the business of the user, just implementation details of the runtime, which will now be built AOT.

File: src/Makefile.in [hunk 1]
```diff
@@ -25,8 +25,10 @@ vpath %.h @srcdir@
 
 #### End of system configuration section. ####
 
+RT_VARIANTS := rt rt_mt rt_sc rt_mt_sc
```
[the remark this answers is not in the document]
Comment (gezalore) [reply src/Makefile.in:28]: I actually started out with the matrix version, but semi arbitrarily dropped them later (too lazy to do error checking in makefiles). As I suggested, we should never need more, so am happy either way.

**Comment by gezalore** [comment]
I pushed a version without the need for the _sc flavour. To do this, all code that might depend on systemc (apart from forward declared types):
- needs to be defined in a header (use `inline` functions)
- the definition must not be available when building the run-time lib (all definitions of `inline` functions must be the same, we will pick up the actual one at link time).

**Comment by wsnyder** [comment]
>I pushed a version without the need for the _sc flavour.

I was afraid this might add a lot of code that then needs to be compiled every header usage, but doesn't seem too bad and is worth the small adder, good idea & thanks.

**Comment by wsnyder** [comment]
>then as part of the build steps, symlink the public headers only under include (and copy them as part of the install step

Note currently you don't presently need to do an install step, that is you can use it directly pointing at VERILATOR_ROOT.  I'd like to retain that.

**Comment by gezalore** [comment]
> Note currently you don't presently need to do an install step, that is you can use it directly pointing at VERILATOR_ROOT. I'd like to retain that.

Yes, that's why we have the symlinks and why the .a is put in lib/ as part of the build. the copy is necessarily for the stand alone install only.

**Comment by gezalore** [comment]
Regarding the ST vs MT build, I did some benchmarking of single threaded model performance:

```
XiangShan --threads 0:

       213,977.83 msec task-clock                #    0.999 CPUs utilized            ( +-  0.01% )
  555,048,316,704      cycles                    #    2.594 GHz                      ( +-  0.01% )
  510,313,918,918      instructions              #    0.92  insn per cycle           ( +-  0.00% )
   49,803,619,652      branches                  #  232.721 M/sec                    ( +-  0.00% )
      456,770,755      branch-misses             #    0.92% of all branches          ( +-  0.60% )

    214.0880 +- 0.0172 seconds time elapsed  ( +-  0.01% )

XiangShan --threads 1:

       213,969.43 msec task-clock                #    0.999 CPUs utilized            ( +-  0.04% )
  555,027,652,335      cycles                    #    2.593 GHz                      ( +-  0.04% )
  510,469,042,865      instructions              #    0.92  insn per cycle           ( +-  0.00% )
   49,831,296,545      branches                  #  232.790 M/sec                    ( +-  0.00% )
      468,616,275      branch-misses             #    0.94% of all branches          ( +-  0.45% )

    214.1433 +- 0.0854 seconds time elapsed  ( +-  0.04% )

OpenTitan --thread 0:

        60,575.65 msec task-clock                #    1.034 CPUs utilized            ( +-  0.05% )
  154,703,794,888      cycles                    #    2.554 GHz                      ( +-  0.05% )
  220,624,742,689      instructions              #    1.43  insn per cycle           ( +-  0.01% )
   19,593,372,934      branches                  #  323.496 M/sec                    ( +-  0.05% )
       75,395,848      branch-misses             #    0.38% of all branches          ( +-  0.13% )

     58.5750 +- 0.0530 seconds time elapsed  ( +-  0.09% )

OpenTitan --threads 1:

        60,754.86 msec task-clock                #    1.034 CPUs utilized            ( +-  0.10% )
  155,177,076,213      cycles                    #    2.552 GHz                      ( +-  0.11% )
  220,697,961,050      instructions              #    1.42  insn per cycle           ( +-  0.01% )
   19,606,571,616      branches                  #  322.499 M/sec                    ( +-  0.05% )
       75,331,140      branch-misses             #    0.38% of all branches          ( +-  0.20% )

     58.7808 +- 0.0607 seconds time elapsed  ( +-  0.10% )
```

So all in all, I don't think that's a difference that warrants the complexity, so we could go with just a single flavour of `libverilated.a` (+ debug version)?

I also did some research on the performance impact of `thread_local`. The crux of it is: if you are statically linking the code using thread local storage, it's basically the same speed as static data:
https://godbolt.org/z/vThW9dTEf

If you need to link the code using thread_local data dynamically, and hence needs to be cimplied with `-fPIC`, you do have a bit of penalty, as you need a function call to resolve the address, though it should still be fast:
https://godbolt.org/z/ra51j3aa4

However, in our current model, we would be linking statically (via libverilated.a), so we can freely use thread_local without having to worry about performance.

What do you think?

**Comment by gezalore** [comment]
Another side note:

With libverilated.a, if the user wants to build a shared library of their generated model, they need to add libverilated.a on the link line of the final executable. 

I.e., previously this used to be enough:
```
gcc -o exe app.o libVmodel.so
```
now they would need this:
```
gcc -o exe app.o libVmodel.so -L$VERILATOR_ROOT -l:libverilated.a
```

Alternatively we could build a PIC version of libverilated that they can `--whole-archive` into their libVmodel.so, but I'm not sure that simplifies things, plus has he potential performance issue with `thread_local`.

**Comment by gezalore** [comment]
@wsnyder let me know what you prefer we do on this, no rush though if you want to think more.

**Comment by wsnyder** [comment]
I think I'm good with the plan, as I understand it:
#. No files move until post-develop-v5 merge (we'll discuss what/where then)
#. Build libverilated and libverilated-dbg, including only mt safe
#. Keep a test like test_regress/t/t_verilated_all that compiles {normal,dbg}_{mt|non-mt} flavors locally versus using the library.

What about gcc vs clang - does everything work if the library is built either way? I realize the ABIs are supposedly compatible but do we need to check no #defines mess that up?

**Comment by wsnyder** [comment]
{edited}

**Comment by gezalore** [comment]
> #. No files move until post-develop-v5 merge (we'll discuss what/where then)

Agreed

> #. Build libverilated and libverilated-dbg, including only mt safe

So 2 versions together debug and non-debug, both with VL_THREADED (MT safe)

> #. Keep a test like test_regress/t/t_verilated_all that compiles {normal,dbg}_{mt|non-mt} flavors locally versus using the library.

Why do we need this still? (I am not sure why that test exists), also why do we want the non-mt version here (apart from prevernting merge pain with master)? I presume after v5 becomes master, VL_THREADED can be removed?

 
> What about gcc vs clang - does everything work if the library is built either way? I realize the ABIs are supposedly compatible but do we need to check no #defines mess that up?

Good point, I will investigate this.

**Comment by wsnyder** [comment]
>Why do we need this still? (I am not sure why that test exists), also why do we want the non-mt version here (apart from prevernting merge pain with master)? I presume after v5 becomes master, VL_THREADED can be removed?

I think we'll end up there eventually, but I don't want to break the single threaded being able to compile until we are sure no one reports performance problems.

**Comment by gezalore** [comment]
> What about gcc vs clang - does everything work if the library is built either way? I realize the ABIs are supposedly compatible but do we need to check no #defines mess that up?

There is a more fundamental problem here, namely GCC using libstdc++ vs Clang using libc++ (by default). With that in mind, I think the net effect of pre-building libverilated.a would be that it's only really safe to build the generated model with the same compiler (and options that affect the ABI) as we used for libverilated.a (i.e.: the one from configure). This of course then means that you would have to have a corresponding verilator installation if you want to change compilers, which we did not need before so this is not ideal.

The reason we actually are attempting to build libverilated.a, is because we had a previous desire for the runtime lib not to be affected by Verilator command line switches or other configs, so multiple models can be mixed soundly (#2359). We have already mostly achieved this in the preparatory patches before this PR (and is unachievable for SC vs non-SC models?).

The runtime being independent of switches then enables using a pre-built archive. This has 2 benefits:
1. Small reduction in build time of model: This is largely irrelevant with ccache, and only matters for small designs (note: the test suite takes virtually the same time to build using libverilated.a vs building from source with ccache).
2. Allow us to reduce the indirect weirdness in the runtime library implementation present due to certain symbols only being available under certain switches (e.g.: the profiler and thread pool setup in the context object). 

The pre-built archive is also useful for testing that the runtime library is indeed independent from the switches.

With that in mind, I would like any alternative solution to satisfy at least:
1. Preserve the runtime lib's independence from the Verilator switches
2. Ensure that the indepencence in point 1 is tested during a test run
3. Provide all symbols in the runtime so we can keep the implementation simpler

I would be OK for one-off build time of the runtime library itself to increase somewhat if ccache can cache it (it's not too long to build it anyway).

**Comment by wsnyder** [comment]
@gezalore I think this closed when develop-v5 was deleted, would you consider rebasing the changes against master?
I'd suggest splitting it into 2 pulls.  First, the C changes needed to just move the functions around, separate from building Verilator, these should be able to work as is without any real side effects.  Then second building upon that the creating the libverilated flavors (note as to above do we need to make separate for gcc + clang?)

**Comment by wsnyder** [comment]
@gezalore github doesn't allow reopening, please file new pulls, thanks.

**Pull request closed without merging**